analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarCompilationStartAnalysisContext.cs

```diff
@@ -42,22 +44,25 @@ public void RegisterNodeAction<TSyntaxKind>(GeneratedCodeRecognizer generatedCod
     {
         if (HasMatchingScope(AnalysisContext.SupportedDiagnostics))
         {
+            ConcurrentDictionary<SyntaxTree, bool> shouldAnalyzeCache = new();
             Context.RegisterSyntaxNodeAction(x =>
-                    Execute<SonarSyntaxNodeReportingContext, SyntaxNodeAnalysisContext>(
-                        new(AnalysisContext, x), action, x.Node.SyntaxTree, generatedCodeRecognizer),
-                syntaxKinds);
-        }
-    }
+            {
+                if (!shouldAnalyzeCache.TryGetValue(x.Node.SyntaxTree, out var canProceedWithAnalysis))
+                {
+                    canProceedWithAnalysis = GetOrAddCanProceedWithAnalysis(generatedCodeRecognizer, shouldAnalyzeCache, x.Node.SyntaxTree);
+                }
+                if (canProceedWithAnalysis)
+                {
+                    action(new(AnalysisContext, x));
+                }
+            }, syntaxKinds);
 
-    /// <inheritdoc cref="SonarAnalysisContext.Execute" />
-    private void Execute<TSonarContext, TRoslynContext>(TSonarContext context, Action<TSonarContext> action, SyntaxTree sourceTree, GeneratedCodeRecognizer generatedCodeRecognizer = null)
-        where TSonarContext : SonarAnalysisContextBase<TRoslynContext>
-    {
-        if (ShouldAnalyzeTree(sourceTree, generatedCodeRecognizer)
-            && SonarAnalysisContext.LegacyIsRegisteredActionEnabled(AnalysisContext.SupportedDiagnostics, sourceTree)
-            && AnalysisContext.ShouldAnalyzeRazorFile(sourceTree))
-        {
-            action(context);
+            // Performance: Don't inline to avoid capture class and delegate allocations.
+            bool GetOrAddCanProceedWithAnalysis(GeneratedCodeRecognizer codeRecognizer, ConcurrentDictionary<SyntaxTree, bool> cache, SyntaxTree tree) =>
+                cache.GetOrAdd(tree, x =>
+                    ShouldAnalyzeTree(x, codeRecognizer)
+                    && SonarAnalysisContext.LegacyIsRegisteredActionEnabled(AnalysisContext.SupportedDiagnostics, x)
+                    && AnalysisContext.ShouldAnalyzeRazorFile(x));
```

Comment:
You better place the method outside of the method. You do not want to create a capture class by accident.